erigon-lib/downloader/mdbx_piece_completion.go

```diff
@@ -118,57 +188,56 @@ func (m *mdbxPieceCompletion) Close() error {
 }
 
 type mdbxPieceCompletionBatch struct {
-	db *mdbx.MdbxKV
+	mdbxPieceCompletion
 }
 
 var _ storage.PieceCompletion = (*mdbxPieceCompletionBatch)(nil)
 
 func NewMdbxPieceCompletionBatch(db kv.RwDB) (ret storage.PieceCompletion, err error) {
-	ret = &mdbxPieceCompletionBatch{db: db.(*mdbx.MdbxKV)}
-	return
-}
-
-func (m *mdbxPieceCompletionBatch) Get(pk metainfo.PieceKey) (cn storage.Completion, err error) {
-	err = m.db.View(context.Background(), func(tx kv.Tx) error {
-		var key [infohash.Size + 4]byte
-		copy(key[:], pk.InfoHash[:])
-		binary.BigEndian.PutUint32(key[infohash.Size:], uint32(pk.Index))
-		cn.Ok = true
-		v, err := tx.GetOne(kv.BittorrentCompletion, key[:])
-		if err != nil {
-			return err
-		}
-		switch string(v) {
-		case complete:
-			cn.Complete = true
-		case incomplete:
-			cn.Complete = false
-		default:
-			cn.Ok = false
-		}
-		return nil
-	})
-	return
+	return &mdbxPieceCompletionBatch{mdbxPieceCompletion{
+		db:        db.(*mdbx.MdbxKV),
+		completed: map[infohash.T]*roaring.Bitmap{}}}, nil
 }
 
 func (m *mdbxPieceCompletionBatch) Set(pk metainfo.PieceKey, b bool) error {
 	if c, err := m.Get(pk); err == nil && c.Ok && c.Complete == b {
 		return nil
 	}
-	var key [infohash.Size + 4]byte
-	copy(key[:], pk.InfoHash[:])
-	binary.BigEndian.PutUint32(key[infohash.Size:], uint32(pk.Index))
 
-	v := []byte(incomplete)
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
 	if b {
-		v = []byte(complete)
+		completed, ok := m.completed[pk.InfoHash]
+
+		if !ok {
+			completed = &roaring.Bitmap{}
+			m.completed[pk.InfoHash] = completed
+		}
+
+		completed.Add(uint32(pk.Index))
+
+		return nil
 	}
+
 	return m.db.Batch(func(tx kv.RwTx) error {
-		return tx.Put(kv.BittorrentCompletion, key[:], v)
+		return putCompletion(tx, pk.InfoHash, uint32(pk.Index), b)
+	})
+}
+
+func (m *mdbxPieceCompletionBatch) Flushed(infoHash infohash.T, flushed *roaring.Bitmap) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	//fmt.Println("FLUSH", infoHash)
+	m.db.Batch(func(tx kv.RwTx) error {
```

Comment:
`Batch` - is designed for much parallel calls. but currently it's behind mutex. In this case `Update` will perform better. 
Or maybe it's possible to do behind mutext only map get/put. 
Or just don't use `mdbxPieceCompletionBatch` if `Update` inside `Set` is not a problem anymore. 
up to you.

Reply:
I've removed this so now we're using the original unbatched version of the code - with the alterations you suggested.